core/src/consensus.rs

```diff
@@ -601,15 +612,23 @@ impl Tower {
                 }
 
                 if last_vote_ancestors.contains(&switch_slot) {
-                    if !self.is_stray_last_vote() {
+                    if self.is_stray_last_vote() {
+                        return suspended_decision_due_to_major_unsynced_ledger();
+                    } else if let Some(latest_duplicate_ancestor) = progress.latest_unconfirmed_duplicate_ancestor(last_voted_slot) {
+                        // We're rolling back because one of the ancestors of the last vote was a duplicate. In this
+                        // case, it's acceptable if the switch candidate is one of ancestors of the previous vote,
+                        // just fail the switch check because there's no point in voting on an ancestor. ReplayStage
+                        // should then have a special case continue building an alternate fork from this ancestor, NOT
+                        // the `last_voted_slot`. This is in contrast to usual SwitchFailure where ReplayStage continues to build blocks
+                        // on latest vote. See `select_vote_and_reset_forks()` for more details.
+                        return rollback_due_to_to_to_duplicate_ancestor(latest_duplicate_ancestor);
```

Comment:
After the rollback, is there any risk that the node will vote on `last_voted_slot` again?

Reply:
so on the initial rollback, a `FailedSwitchDuplicateRollback` will be triggered here, so the `switch_fork_decision.can_vote()` will return false, and the votable bank should be returned as `None` here: https://github.com/solana-labs/solana/blob/master/core/src/replay_stage.rs#L1677-L1689. Note the `!is_locked_out` check here as well. Future banks will incur this check https://github.com/solana-labs/solana/blob/master/core/src/replay_stage.rs#L1528 (not here I think even after purging the duplicate block from the ancestors/descendants/bank forks/progress, when the ancestors are removed (which are necessary for calculating that `is_locked_out()` check), that `is_locked_out` will return false. However after we repair the correct version of the slot, I think this check: https://github.com/solana-labs/solana/blob/master/core/src/consensus.rs#L490 will need to be adjusted to account for a hash or a flag needs to say the current version of slot `N` is different than the last voted version of slot `N`, so lockout still needs to be applied